orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/support/TerminatingInstanceSupport.groovy

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2015 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.clouddriver.pipeline.support
+
+import com.netflix.spinnaker.orca.clouddriver.utils.CloudProviderAware
+import com.netflix.spinnaker.orca.clouddriver.utils.OortHelper
+import com.netflix.spinnaker.orca.pipeline.model.Stage
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+import retrofit.RetrofitError
+
+@Slf4j
+@Component
+class TerminatingInstanceSupport implements CloudProviderAware {
+
+  /**
+   * Key for each iteration of a task to check on the progress of termination for each instance. Value is a list
+   * of TerminatingInstance objects.
+   */
+  static final String TERMINATE_REMAINING_INSTANCES = "terminate.remaining.instances"
+
+  @Autowired
+  OortHelper oortHelper
+
+  /**
+   * Some platforms kill instances within server groups and recreate them with different IDs - for these, we just
+   * ensure the instance specified no longer exists.
+   * Other platforms  kill instances within server groups and recreate them with the same ID - for these, we can't
+   * just check for the absence of the instance alone. Additionally, we must compare the launch times of the new
+   * instance with that of the killed one.
+   * For instances that are not part of any server group (stand alone instances), we ensure the instance no longer
+   * exists and do not expect it to be recreated.
+   */
+  List<TerminatingInstance> remainingInstances(Stage stage) {
+    List<TerminatingInstance> terminatingInstances
+    if (stage.context[TERMINATE_REMAINING_INSTANCES]) {
+      terminatingInstances = Arrays.asList(stage.mapTo("/" + TERMINATE_REMAINING_INSTANCES, TerminatingInstance[]))
+    } else {
+      List<String> instanceIds = stage.context.instanceIds // Used by terminateInstances.
+      if (!instanceIds) {
+        String instanceId = stage.context.instance // Used by terminateInstanceAndDecrementServerGroup.
```

Comment:
we need to do this ( https://github.com/spinnaker/clouddriver/pull/235 ) in orca at some point too...